### PR TITLE
Exception handling for macOS

### PR DESCRIPTION
**Pull request by lptr**
Fixes when merged via `gradle/gradle`: https://github.com/gradle/gradle/issues/12219

### CODE REVIEW

File: src/main/cpp/apple_fsnotifier.cpp [hunk 1]
```diff
@@ -222,56 +226,65 @@ JNIEnv* Server::getThreadEnv() {
...
     for (int i = 0; i < count; i++) {
         jstring path = (jstring) env->GetObjectArrayElement(paths, i);
-        char* watchedPath = java_to_char(env, path, result);
-        log_fine(env, "Watching %s", watchedPath);
+        char* watchedPath = java_to_char(env, path, NULL);
         if (watchedPath == NULL) {
-            log_severe(env, "Could not allocate string to store root to watch.", NULL);
-            // TODO Free resources
-            return NULL;
+            throw FileWatcherException("Could not allocate string to store root to watch");
```
Comment (wolfs) [inline src/main/cpp/apple_fsnotifier.cpp]: Don't we need to free resources when the exception is thrown?

Comment (lptr) [reply src/main/cpp/apple_fsnotifier.cpp]: That part is not done yet. I am actively learning about `unique_ptr` and friends to make sure they get released.

**Comment by lptr** [comment]
@wolfs Yeah, I'm also unsure as to how to test error handling. :( Let me know if you have ideas. Meanwhile PTAL.

File: src/main/cpp/apple_fsnotifier.cpp (hunk 1 above)
Comment (lptr) [reply src/main/cpp/apple_fsnotifier.cpp]: I think we are actually much better now wrt freeing resources when a problem occurs.

File: src/main/cpp/apple_fsnotifier.cpp [hunk 2]
```diff
@@ -7,6 +7,20 @@
 
 using namespace std;
 
+struct FileWatcherException : public exception {
```
Comment (wolfs) [inline src/main/cpp/apple_fsnotifier.cpp]: I suppose this should move to `generic.cpp` at some point, right?

File: src/main/cpp/apple_fsnotifier.cpp [hunk 3]
```diff
@@ -7,6 +7,20 @@
...
+public:
+    FileWatcherException(const char *message) {
+        this->message = message;
+    }
+
+    const char * what () const throw () {
```
Comment (wolfs) [inline src/main/cpp/apple_fsnotifier.cpp]:
```suggestion
    const char* what() const throw() {
```

Nit. What code style shall we use here? Can we maybe configure VSCode to autoformat the native code?

File: src/main/cpp/apple_fsnotifier.cpp [hunk 4]
```diff
@@ -85,39 +103,39 @@ Server::Server(JNIEnv *env, jobject watcherCallback, CFArrayRef rootsToWatch, lo
...
-        log_severe(env, "Could not create FSEventStreamCreate to track changes", NULL);
-        // TODO Error handling
-        return;
+        throw FileWatcherException("Could not create FSEventStreamCreate to track changes");
     }
     this->watcherStream = watcherStream;
+
+    unique_lock<mutex> lock(watcherThreadMutex);
     this->watcherThread = thread(&Server::run, this);
+    this->watcherThreadStarted.wait(lock);
```
Comment (wolfs) [inline src/main/cpp/apple_fsnotifier.cpp]: Maybe we shouldn't wait forever here. Is that possible?

**Pull request merged**